src/commands.cpp

```diff
@@ -324,6 +324,33 @@ CommandPutFile::CommandPutFile(MegaClient* client, TransferSlot* ctslot, int ms)
     arg("v", 2);
     arg("s", tslot->fa->size);
     arg("ms", ms);
+
+    // send minimum set of different tree's roots for API to check overquota
+    set<handle> targetRoots;
+    beginarray("t");
+    for (auto &file : tslot->transfer->files)
+    {
+        if (!ISUNDEF(file->h))
+        {
+
+            Node *node = client->nodebyhandle(file->h);
+            if (node)
+            {
+                handle rootnode = client->getrootnode(node)->nodehandle;
+                tslot->transfer->mForeignTarget = client->rootnodes[0] != rootnode;
```

Comment:
mForeignTarget will be set according to the last processed `file` in the list, without taking into account that other files in the list might be 'foreign' or not.   Wouldn't it be better to set it according to any-are-foreign or all-are-foreign so that the order of the files in the list does not matter?   thanks

Reply:
You're right, @mattw-mega . I thought it was managed on a per-`File` basis, rather than per `Transfer`. Since we want to adjust the behavior of failed transfers for `u` with -17 based on the set of target handles, we do need to know if all-are-foreign, if there are mixed foreign-plus-own handles or all were own handles at the time the `u` is sent, and do not retry transfers with foreign target if the `u` failed with -17. If there was a mix, then keep retrying the ones with targets in you own account, but permanent fail the foreign.